### PR TITLE
[Gluten-2600][CH] fix client cache concurrent issue

### PR DESCRIPTION
**Pull request by binmahone**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#ISSUE-ID)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp [hunk 1]
```diff
@@ -69,6 +71,34 @@ namespace ErrorCodes
 
 namespace local_engine
 {
+template <class key_type, class value_type>
```
Comment (zzcclp) [inline cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp:74]: put this class in package Common?

File: cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp [hunk 2]
```diff
@@ -496,16 +527,20 @@ class S3FileReadBufferBuilder : public ReadBufferBuilder
         if (!getSetting(settings, bucket_name, BackendInitializerUtil::HADOOP_S3_ASSUMED_ROLE, "", true).empty())
             is_per_bucket = true;
 
-        if (is_per_bucket && per_bucket_clients.contains(bucket_name)
-            && "true" != getSetting(settings, bucket_name, BackendInitializerUtil::HADOOP_S3_CLIENT_CACHE_IGNORE))
+        if (is_per_bucket && "true" != getSetting(settings, bucket_name, BackendInitializerUtil::HADOOP_S3_CLIENT_CACHE_IGNORE))
         {
-            return per_bucket_clients.get(bucket_name).get();
+            auto client = per_bucket_clients.get(bucket_name);
+            if (client.has_value())
```
Comment (zzcclp) [inline cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp:533]: when the client is 'boost::none', can it call has_value() function?

File: cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp (hunk 1 above)
Comment (binmahone) [reply cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp:74]: temporally put here, can be moved to common when reuse scenario occurs

File: cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp (hunk 2 above)
Comment (binmahone) [reply cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp:533]: if it's none, has_value should return false?

Comment (zzcclp) [reply cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp:533]:
```
    boost::optional<size_t> test_optial = boost::none;
    boost::optional<size_t> aaa = 23;
    std::cout << test_optial.has_value() << std::endl;
    std::cout << aaa.has_value() << std::endl;
```
test code, it works.

File: cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp (hunk 1 above)
Comment (zzcclp) [reply cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp:74]: OK

**Review by zzcclp — APPROVED** [review]
LGTM

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**